lib/internal/modules/esm/get_format.js

```diff
@@ -135,6 +135,52 @@ function getFileProtocolModuleFormat(url, context = { __proto__: null }, ignoreE
       }
     }
   }
+  const tsEnabled = getOptionValue('--experimental-strip-types');
+  if (ext === '.ts' && tsEnabled) {
+    const { type: packageType, pjsonPath } = getPackageScopeConfig(url);
+    if (packageType !== 'none') {
+      return `${packageType}-typescript`;
+    }
+    // The controlling `package.json` file has no `type` field.
+    switch (getOptionValue('--experimental-default-type')) {
+      case 'module': { // The user explicitly passed `--experimental-default-type=module`.
+        // An exception to the type flag making ESM the default everywhere is that package scopes under `node_modules`
+        // should retain the assumption that a lack of a `type` field means CommonJS.
+        return underNodeModules(url) ? 'commonjs-typescript' : 'module-typescript';
+      }
+      case 'commonjs': { // The user explicitly passed `--experimental-default-type=commonjs`.
+        return 'commonjs-typescript';
+      }
+      default: { // The user did not pass `--experimental-default-type`.
+        // `source` is undefined when this is called from `defaultResolve`;
+        // but this gets called again from `defaultLoad`/`defaultLoadSync`.
+        if (getOptionValue('--experimental-detect-module')) {
+          if (tsEnabled) {
+            const { tsParse } = require('internal/modules/helpers');
+            source = tsParse(source);
+          }
+          const format = source ?
+            (containsModuleSyntax(`${source}`, fileURLToPath(url), url) ? 'module-typescript' : 'commonjs-typescript') :
+            null;
+          if (format === 'module-typescript') {
+            // This module has a .ts extension, a package.json with no `type` field, and ESM syntax.
+            // Warn about the missing `type` field so that the user can avoid the performance penalty of detection.
+            typelessPackageJsonFilesWarnedAbout ??= new SafeSet();
+            if (!typelessPackageJsonFilesWarnedAbout.has(pjsonPath)) {
+              const warning = `${url} parsed as an ES module because module syntax was detected;` +
+                ` to avoid the performance penalty of syntax detection, add "type": "module" to ${pjsonPath}`;
+              process.emitWarning(warning, {
+                code: 'MODULE_TYPELESS_PACKAGE_JSON',
+              });
+              typelessPackageJsonFilesWarnedAbout.add(pjsonPath);
+            }
+          }
+          return format;
+        }
+        return 'commonjs-typescript';
```

Comment:
This seems all copy-pasted from how we handle `.js` files. Can we avoid the copy-paste?

Reply:
its similar but not copy pasted